[no squash] Split up tile.cpp/tile.h and just a little refactoring

cx384 opened the pull request:
## Goal of the PR
Improve the maintainability of `tile.cpp` and potential compilation time improvements.
Lately there have been many issues which relate to this file, to name a few:
- Bugs: #14223 #14241 #14242 #14312 #14270 #12459
- Performance: #14322 #11531 #13998
- Features: #11588 #13501 #9946

The first step, to refactor it, is to split it into its components, such that only the needed parts have to be included.
This also makes reviewing and merging of future PRs easier.

## How does the PR work?
- `tile.(h,cpp)` gets split into three separate parts, which are all completely independent and don't depend on each other:
  1. **`texture_paths.(h,cpp)`** contains the texture path functions. Moving it out from `tile.h` was already marked as TODO in the old file. Moreover, `texture_paths.cpp` did undergo some obvious **refactoring**.
  3. **`texturesource.(h,cpp)`** contains the `ITextureSource` class, which is the main reason why the old `tile.h` got included. These files got **not refactored** and contain exactly the same as the old files (only some includes changed), since it would be too much for this PR. So you don't need to test them.
  4. **`tile.(h,cpp)`** contains anything that has actually something to do with tiles, e.g. `TileSpec` or `TileLayer`. Of this part also nothing changed, except for de-inlineing two functions.
 
- The `Align2Npot2` function has moved to `imagefilters.h`, since it doesn't fit into the other files.
  (Alternatively, it could become a member of the ITextureSource class.)
- A lot of includes have changed all across the code base. I tried to remove all unnecessary ones and forward declare if possible. As long as it compiles, it should be fine. If anything is needed, it can be included again.
- I kept every commit, which just moves things, separate, such that it can be easier reviewed. It should be buildable after any commit.


## Does this relate to a goal in [the roadmap](https://github.com/minetest/minetest/blob/master/doc/direction.md)?
- 2.2 Internal code refactoring

## To do

This PR Ready for Review.

(`texturesource.cpp` needs to be refactored. I already started this, but it would be too much for this PR.
Maybe it should even be split into more files, but it wouldn't make a difference in the header.)


## How to test

- Notice that it compiles.
- Read the code.
- Use `diff <old tile.cpp file> <new texturesource.cpp file>` or similar to see that indeed nothing of the `ITextureSource` class has changed. The other things can be compared by hand.
- See that the refactored `texture_paths.cpp` still works (should be obvious).

sfan5 commented:
Not a lot left in tile.cpp after this but I think the split makes sense. :+1: 

One thing: considering our existing naming scheme `texturepaths.cpp` is more fitting than `texture_paths.cpp`.

What is also important is that the changes that actually change code are cleanly split from moving code. So in a mergeable state I think this PR should contain two commits:
* Split up tile.cpp/h
* Refactoring

sfan5 commented on src/client/texture_paths.cpp:
```diff
@@ -0,0 +1,108 @@
+		if (fs::PathExists(extended_path))
+			return extended_path;
+	}
+	while(*(++ext) != NULL);
```
a `for (auto &ext : extensions)` would be more C++-y, just be careful of the null value

cx384 replied:
Ok, I modernized it.
Note that it now uses a vector instead of a C style array, since as far as I know, C++17 does not have `ranges::subrange` yet and I don't think you can limit the size of a C++ array without using additional memory.

cx384 commented on src/client/texture_paths.cpp:
```diff
@@ -0,0 +1,108 @@
+std::string getImagePath(std::string_view path)
+{
+	// A NULL-ended list of possible image extensions
+	const char *extensions[] = {".png", ".jpg", ".bmp", ".tga", NULL};
```
I kept this intentionally since the description of `removeStringEnd` talks about NULL.
But now it is a `nullptr`.

cx384 commented:
I modernized the things you suggested,
renamed `texture_paths.h` to `texturepaths.h`,
edited the git history to only contain a split commit and a refactoring commit.
and rebased.

sfan5 commented on src/client/texturepaths.cpp:
```diff
@@ -0,0 +1,113 @@
+	// A NULL-ended list of possible image extensions
+	// (In newer C++ versions a fixed size array and ranges::subrange could be used
+	// or just modernise removeStringEnd.)
+	std::vector<const char*> extensions{".png", ".jpg", ".bmp", ".tga", nullptr};
+
+	// Remove present extension
+	std::string_view stripped_path = removeStringEnd(path, extensions.data());
+	// If there is no known extension, assume it has been omitted.
+	if (stripped_path.empty())
+		stripped_path = path;
+
+	extensions.pop_back();
+	for (const char *ext : extensions) {
+		std::string extended_path{stripped_path};
+		extended_path.append(ext);
+		if (fs::PathExists(extended_path))
+			return extended_path;
+	}
```
surely this would have been simpler?
```suggestion
	const char *extensions[] = {".png", ".jpg", ".bmp", ".tga", nullptr};

	// Remove present extension
	std::string_view stripped_path = removeStringEnd(path, extensions);
	// If there is no known extension, assume it has been omitted.
	if (stripped_path.empty())
		stripped_path = path;

	for (const char *ext : extensions) {
		if (!ext)
			break;
		std::string extended_path(stripped_path);
		extended_path.append(ext);
		if (fs::PathExists(extended_path))
			return extended_path;
	}
```

cx384 replied:
Unnecessarily calling `if (!ext)` 5 times for every path does not look very nice imo. (even though it will probably be optimized away)

Ideally, `removeStringEnd` should be changed to take a vector as parameter instead of a NULL terminated C array, I think.

Do you want me to apply your suggestion, or can you live with my code?

sfan5 reviewed (approved):
works

cx384 commented:
Rebased again.

Desour reviewed (approved):
LGTM. :+1:
